apps/els_lsp/src/els_refactorerl_utils.erl

```diff
@@ -0,0 +1,284 @@
+
+%%==============================================================================
+%% API
+%%==============================================================================
+
+%% @doc
+%% Returns the RefactorErl node, if it can't, it returns error and its cause.
+%% It returns the node given in config, if it is alive.
+%% First it runs the validation functions, the result of validation will be
+%% notified to the user.
+%% If the node once was validated there will be no display messages.
+%%
+%% The configuration can store the node and its status.
+%%  - Either a simple NodeString, which needs to be checked and configure
+%%  - {Node, Status} where boths are atoms
+%%      Possible statuses: validated, disconnected, disabled
+%%      'disabled',       when it won't try to reconnect
+%%      'disconnected',   when it will try to reconnect
+%%  - notconfigured, the node is not configured in the config file
+%%
+%%
+%% Node can be:
+%%  - NodeStr
+%%  - {Status, Node} where both are atoms.
+%%  - Status can be:
+%%    - validated:    node is running
+%%    - disconnected: node is not running
+%%  - disabled: RefactorErl is turned off for this session. T
+%%              his can happen after an unsuccessfull query attempt.
+-spec referl_node() -> {ok, atom()}
+                     | {error, disconnected}
+                     | {error, disabled}
+                     | {error, other}.
+referl_node() -> 
+  case els_config:get(refactorerl) of
+    #{"node" := {Node, validated}} ->
+      {ok, Node};
+
+    #{"node" := {Node, disconnected}} ->
+      connect_node({retry, Node});
+
+    #{"node" := {_Node, disabled}} ->
+      {error, disabled};
+
+    #{"node" := NodeStr} ->
+      RT = els_config_runtime:get_name_type(),
+      Node = els_utils:compose_node_name(NodeStr, RT),
+      connect_node({validate, Node});
+
+    notconfigured ->
+      notification("RefactorErl is not configured!"),
+      {error, disabled};
+
+    _ ->
+      {error, other}
+  end.
+
+
+%%@doc
+%% Adds a module to the RefactorErl node. Using the UI router
+%% Returns 'ok' if successfull
+%% Returns 'error' if it fails after ?MAX_RECURSION_DEPTH number of tries
+-spec add(uri()) -> atom().
+add(Uri) -> 
+  add(Uri, 0).
+
+-spec add(uri(), number()) -> atom().
+add(Uri, RecursionDepth) when RecursionDepth < ?MAX_RECURSION_DEPTH ->
+  case els_refactorerl_utils:referl_node() of
+    {ok, Node} ->
+      Path = [binary_to_list(els_uri:path(Uri))],
+      ReqID = request_id(),
+      Resp = rpc:call(  Node
+                      , reflib_ui_router
+                      , request
+                      , [self(), ReqID, {add_dir, Path}] ),
+      case Resp of
+        {badrpc, _} ->
+          disable_node(Node),
+          error;
+        ok ->
+          receive
+            {ReqID, reply, {ok, _}} -> ok
+          end;
+        deny ->
+          notification("Adding is deined, retry!"), % TODO: Take this out overtime stability proves itself
+          timer:sleep(1000),
```

Comment:
I would really avoid retries. 1 second during editing is an extremely significant delay and the user experience would be seriously degraded. We want to keep the Erlang LS side of the integration as simple as possible (we will have to maintain it, after all), so every other instability issue should be solved on the RefactorErl side in my mind. In case of any error, the correct thing to do in my mind is not to report any diagnostics (`[]`). So, to be able to merge this, I would really encourage you to take away all the error handling, at least until we get the basic integration up and running.